code/modules/biomatter_manipulation/biogenerator.dm

```diff
@@ -0,0 +1,494 @@
+
+
+//Power biomatter generator
+//This machine use biomatter reagent and some of O2 to produce power (it also produce CO2)
+//It has a few components that can be weared out, so operator should check this machine from time o time and tinker it
+//In this case, our multistructure datum not just holder, but core of our machine and process by its own
+
+#define WEAROUT_CHANCE 10
+
+/datum/multistructure/biogenerator
+	structure = list(
+		list(/obj/machinery/multistructure/biogenerator_part/generator, /obj/machinery/multistructure/biogenerator_part/console, /obj/machinery/multistructure/biogenerator_part/port)
+					)
+	var/obj/machinery/multistructure/biogenerator_part/console/screen
+	var/obj/machinery/multistructure/biogenerator_part/port/port
+	var/obj/machinery/multistructure/biogenerator_part/generator/generator
+
+	var/working = FALSE
+	var/last_output_power = 0		//used at UI
+
+
+/datum/multistructure/biogenerator/init()
+	..()
+	START_PROCESSING(SSprocessing, src)
+
+
+/datum/multistructure/biogenerator/Destroy()
+	STOP_PROCESSING(SSprocessing, src)
+	return ..()
+
+
+/datum/multistructure/biogenerator/connect_elements()
+	..()
+	screen 		= locate() in elements
+	port 		= locate() in elements
+	generator	= locate() in elements
+
+
+/datum/multistructure/biogenerator/is_operational()
+	. = ..()
+	if(!.)
+		return FALSE
+
+	if(!generator.core || !generator.chamber)
```

Comment:
if `generator = locate()` fucks up, won't `!generator.core` generate runtime error becouse of null object ref?

Reply:
It's okay. This proc will not fire if at least one of parts is missing.
But sure, i'll add a handle for such case